ran-core/src/test/java/io/ran/ClazzMethodTest.java

```diff
@@ -0,0 +1,535 @@
+		// ListHolder<List<String>> make(); ListHolder<F extends List<?>>
+		// needs to decide between ListHolder<List<String>> and ListHolder<List<Object>>
+		ClazzMethod makeMethod = Clazz.of(ListHolderFactoryOfString.class).methods().find("make", ListHolder.class).get();
+		assertMethod(ListHolderFactoryOfString.class, makeMethod, g(ListHolder.class, g(List.class, String.class)));
+		assertFalse(makeMethod.hasGenericFromClass());
+		assertFalse(makeMethod.hasGenericFromMethod());
+
+		ClazzMethod getMethod = makeMethod.getReturnType().methods().find("get", List.class).get();
+		assertMethod(g(ListHolder.class, g(List.class, String.class)), getMethod, g(List.class, String.class));
+		assertTrue(getMethod.hasGenericFromClass());
+		assertFalse(getMethod.hasGenericFromMethod());
+	}
+
+	@Ignore("not supported atm")
+	@Test
+	public void testMoreSpecificFirstLevelSame2() {
+		// StringListHolder<? extends List<?>> make(); StringListHolder<F extends List<String>>
+		// needs to decide between StringListHolder<List<Object>> or StringListHolder<List<String>>
+		ClazzMethod makeMethod = Clazz.of(StringListHolderFactoryOfWildcard.class).methods().find("make", StringListHolder.class).get();
+		assertMethod(StringListHolderFactoryOfWildcard.class, makeMethod, g(StringListHolder.class, g(List.class, String.class)));
+		assertFalse(makeMethod.hasGenericFromClass());
+		assertTrue(makeMethod.hasGenericFromMethod());
+
+		ClazzMethod getMethod = makeMethod.getReturnType().methods().find("get", List.class).get();
+		assertMethod(g(StringListHolder.class, g(List.class, String.class)), getMethod, g(List.class, String.class));
+		assertTrue(getMethod.hasGenericFromClass());
+		assertFalse(getMethod.hasGenericFromMethod());
+	}
+
+	@Ignore("not supported atm")
+	@Test
+	public void testMoreSpecificEachLevelDifferent() {
+		// ListHolder<? extends Collection<String>> make(); ListHolder<F extends List<?>>
+		// needs to decide between ListHolder<List<Object>> or ListHolder<Collection<String>>
+		ClazzMethod makeMethod = Clazz.of(StringCollectionHolderFactory.class).methods().find("make", ListHolder.class).get();
+		assertMethod(StringCollectionHolderFactory.class, makeMethod, g(ListHolder.class, g(List.class, String.class)));
+		assertFalse(makeMethod.hasGenericFromClass());
+		assertTrue(makeMethod.hasGenericFromMethod());
+
+		ClazzMethod getMethod = makeMethod.getReturnType().methods().find("get", List.class).get();
+		assertMethod(g(ListHolder.class, g(List.class, String.class)), getMethod, g(List.class, String.class));
+		assertTrue(getMethod.hasGenericFromClass());
+		assertFalse(getMethod.hasGenericFromMethod());
+	}
+
+	@Test
+	public void testMoreSpecificDifferentBounds() {
+		// ListHolder<? extends Set<String>> make(); ListHolder<F extends List<?>>
+		// this is effectively multiple bounds, not supported
+		assertThrows(IllegalArgumentException.class,
+				() -> Clazz.of(StringSetHolderFactory.class).methods().find("make", ListHolder.class));
```

Comment:
Not sure Iunderstand why this is an illegal operation

Reply:
it is effectively multiple bound generics, which we don't support. The method returns ListHolder of something that is both a Set<String> and List<?>